temporal-opentracing/src/main/java/io/temporal/opentracing/OperationNameAndTagsProvider.java

```diff
@@ -0,0 +1,42 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.opentracing;
+
+import java.util.Map;
+
+/** Allows a pluggable mechanism to configure an OpenTracing span's name and tags */
+public interface OperationNameAndTagsProvider {
```

Comment:
I suggest we turn this around a little bit and create pluggable `SpanContextBuilder` and make these functions `getSpanName` and `getSpanTags` available on the span context itself. This way we won't need this new abstraction that extracts tags/name out of the span context.

Reply:
I'm a little unsure of what you're suggesting here.  Are you suggesting that we have a `StartSpanContextBuilder` that is set into `OpenTracingOptions` that has methods like `setWorkflowId` and `setRunId` for `SpanFactory` to call (as per your other comment), and also has implementations for `getSpanName` and `getSpanTags` that can be overridden as needed?

Reply:
I'm not in agreement with this proposal. 

> pluggable SpanContextBuilder

Builder pattern is not a strategy/provider/factory pattern. Builder shouldn't be pluggable. Builder shouldn't provide different implementations. We can use Builder pattern to just keep span name and tag in one basket if we want, but we shouldn't make Builder pluggable.

> getSpanName and getSpanTags available on the span context itself.

That's not what we want to achieve here. We want the same span context to produce different span names and tags depending on the plugged naming strategy.

> This way we won't need this new abstraction that extracts tags/name out of the span context.

I think that the creation of such abstraction is the sole goal of this PR...

Reply:
Ok, I think I maybe got what Vitaly proposes.

We have SpanFactory#createSpan and we can refactor it to call something like:
```
  interface SpanBuilderProvider {
    Tracer.SpanBuilder createSpanBuilder(Tracer tracer, StartSpanContext startSpanContext);
  }
```
and the implementation will be responsible for a population of information from StartSpanContext to Tracer.SpanBuilder. After that SpanFactory can finishing publishing of additional information into this builder.


and we can have a default implementation that will use our new OperationNameAndTagsProvider:
```
  class SpanBuilderFromSpanContentProvider {
    private final OperationNameAndTagsProvider spanContentProvider;

    public SpanBuilderFromSpanContentProvider(OperationNameAndTagsProvider spanContentProvider) {
      this.spanContentProvider = spanContentProvider;
    }

    Tracer.SpanBuilder createSpanBuilder(Tracer tracer, StartSpanContext context) {
      Tracer.SpanBuilder spanBuilder =
              tracer.buildSpan(spanContentProvider.getOperationName(context));


      spanContentProvider.getTags(context).forEach(spanBuilder::withTag);

      return spanBuilder;
    }
  }
```

And we can in that case define an instance of SpanBuilderProvider on OpenTracingOptions to give full flexibility to the span creation process. Not just methods getName and getTags, but this way we can fully hijack the process if we want.

I'm not sure. From one point of view, this way we make things even more flexible, but I'm not a big fan. I think it's a more clean way for temporal users to have a well defined interface with simpler methods like getName and getTags. I will not opposite this solution, but I think it's an overkill here. Maybe we can implement it only if somebody really needs it for something.

Reply:
Yes, my proposal is to put tags/name into span context and have a builder that builds it based on all available parameters. This way you could make default implementation and also allow provide ability to supply a custom builder, which would allow any type of transformation from parameters to tags/name. 
Functionally it will be the same, but will provide a little cleaner abstractions allowing us to get rid of 'OperationNameAndTagsProvider`.

Reply:
I like Scott's proposal with an interface with clearly defined `#getSmth` methods because it makes the implementation agnostic of OpenTracing Java classes. It will be a smoother design if we implement OpenTelemetry natively in addition to OpenTracing. Which is effectively the same as OpenTracing, but has different classes structure / packages. 
The current OperationNameAndTagsProvider interface and user's implementations of it can be used for both OpenTelemetry and OpenTracing without any code changes.
So, I want to propose that if we support a fully flexible SpanBuilderFromSpanContentProvider#createSpanBuilder - the default implementation of it should use OperationNameAndTagsProvider and accept it as a constructor parameter. This way majority of users will use our default SpanBuilderFromSpanContentProvider and provide only a custom OperationNameAndTagsProvider. And users who want full access to the span can override the whole SpanBuilderFromSpanContentProvider.

Reply:
OK, I wound up using the SpanContentProvider concept without the additional OperationNameAndTags provider - it just seemed a little cleaner and it's really easy to just override the `getSpanName` and `getSpanTags` on the default implementation.  Please take a look